### PR TITLE
Add new `redistribute()` directive

### PR DESCRIPTION
**Pull request by scheibelp**
Some packages can't be redistributed in source or binary form. We need an explicit way to say that in a package.

This adds a `redistribute()` directive so that package authors can write, e.g.:

```python
    redistribute(source=False, binary=False)
```

You can also do this conditionally with `when=`, as with other directives, e.g.:

```python
    # 12.0 and higher are proprietary
    redistribute(source=False, binary=False, when="@12.0:")

    # can't redistribute when we depend on some proprietary dependency
    redistribute(source=False, binary=False, when="^proprietary-dependency")
```

To prevent Spack from adding either their sources or binaries to public mirrors and build caches. You can still unconditionally add things *if* you run either:
* `spack mirror create --private`
* `spack buildcache push --private`

But the default behavior for build caches is not to include non-redistributable packages in either mirrors or build caches.  We have previously done this manually for our public buildcache, but with this we can start maintaining redistributability directly in packages.

Caveats: currently the default for `redistribute()` is `True` for both `source` and `binary`, and you can only set either of them to `False` via this directive.

- [x] add `redistribute()` directive
- [x] add `redistribute_source` and `redistribute_binary` class methods to `PackageBase`
- [x] add `--private` option to `spack mirror`
- [x] add `--private` option to `spack buildcache push`
- [x] test exclusion of packages from source mirror (both as a root and as a dependency)
- [x] test exclusion of packages from binary mirror (both as a root and as a dependency)

### CODE REVIEW

**Comment by samcmill** [comment]
The approach looks goods to me.

**Comment by tgamblin** [comment]
~~Per our discussion, I've converted this to draft until @scottwittenburg comes back from break -- we want to make sure pipeline folks are around before we merge to develop~~

**Comment by tgamblin** [comment]
oops I converted the wrong PR to draft -- false alarm sorry

**Comment by rscohn2** [comment]
@scheibelp: It would be useful to have this functionality. Is there something I can do to help?

**Comment by tgamblin** [comment]
@rscohn2: we recently added #39346, and I talked to @boomanaiden154 about having something like [this file](https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix) at the top-level in Spack, describing what each license allows (distributing sources, binaries, etc.). Given that we would probably try to loop this functionality into the `license()` directive if possible.

Is that something someone at Intel would like to help with?  Or maybe if @boomanaiden154 has made more progress, he already has something.

**Comment by boomanaiden154** [comment]
I have not made any progress on getting that implemented (been busy with school and other projects). I also haven't given any thought/effort to using the license information to restrict pulling in packages/what gets pushed to a mirror, but it seems like a compelling use case of the license directive. Doing that would still require implementing an SPDX parser, but that should be relatively simple, it just needs doing.

**Comment by scheibelp** [comment]
> having something like [this file](https://github.com/NixOS/nixpkgs/blob/master/lib/licenses.nix) at the top-level in Spack, describing what each license allows (distributing sources, binaries, etc.)

* Presumably there is no reason for preventing this other than licensing. However, if there are other cases, I'll note that this PR was general enough to handle them (whereas encoding this into the licensing file would be problematic)
* Our `license` declaration requires an SPDX identifier ([I think](https://spack.readthedocs.io/en/latest/packaging_guide.html#specifying-license-information)). It may not be clear to users how to create their own (and there may be drift between what the decide here and what they submit to the SPDX database). I'll note the file you linked from `nixpkgs` appears to be able to accommodate non-SPDX licenses
* I think users approaching this from the use case of wanting to prevent downloads might find the license approach more indirect (create a new license, update the license file to specify that instances of packages with that license should not be redistributed)
  * Furthermore there seems to be potential for silent failures if the user mistypes a license identifier (which can either miss entirely or pick an existing-but-wrong license)

**Comment by boomanaiden154** [comment]
>Presumably there is no reason for preventing this other than licensing. However, if there are other cases, I'll note that this PR was general enough to handle them (whereas encoding this into the licensing file would be problematic)

Are there other cases that exist currently?

>Our license declaration requires an SPDX identifier ([I think](https://spack.readthedocs.io/en/latest/packaging_guide.html#specifying-license-information)). It may not be clear to users how to create their own (and there may be drift between what the decide here and what they submit to the SPDX database). I'll note the file you linked from nixpkgs appears to be able to accommodate non-SPDX licenses

It is intended to require an SPDX identifier, but that isn't enforced currently (still need to get to implementing that in `spack audit`). The idea was that all the identifiers would be listed in the config file, and items could be added as needed, coordinating with upstream SPDX to see that we're consistent. I don't have much experience working with upstream SPDX on the integration of new licenses, but the team working on that seems reasonable to work with.

>I think users approaching this from the use case of wanting to prevent downloads might find the license approach more indirect (create a new license, update the license file to specify that instances of packages with that license should not be redistributed)
Furthermore there seems to be potential for silent failures if the user mistypes a license identifier (which can either miss entirely or pick an existing-but-wrong license)

Presumably the license would either be a standard license already identified by SPDX (most likely), or proprietary software that probably doesn't allow redistribution anyways? We haven't exactly figured it out yet, but I think the plan was to have a custom `PROPRIETARY` (or some equivalent) identifier that could be used for that type of software. Silent failures are possible right now, but something that would eventually be resolved by adding license validation into `spack audit`, ensuring that the license for a package maps to something.

**Comment by rscohn2** [comment]
I agree with @scheibelp's concerns.

* I like the way that this PR lets me add an attribute to a package to prevent redistribution. It is easy to see and check. I don't have to worry that some change to a license database somewhere will change behavior for my package in an unintended way.
* For the intel packages, I don't want to be required to provide a SPDX identifier on even to provide the license inline. I just want to put a link to license information. This is what intel asked me to do for all the intel proprietary packages (see `spack info intel-oneapi-compilers`).
* I have general concerns about spack package maintainers interpreting licenses for the package. They are often not maintainers for the upstream project. Some projects have simple licensing, but many do not. It is hard to keep up to date.

File: lib/spack/spack/cmd/buildcache.py
Comment (tgamblin) [inline lib/spack/spack/cmd/buildcache.py]: I think we should instead default to `public` and allow a flag for `--private`.

I know that *most* build caches are likely going to be private but I think you should have to be explicit to indicate that you're only making this for internal purposes, and that you're ok with distributing otherwise un-distributable packages as long as it's internal.

Making `--public` the special case seems like an easy way for people to distribute things inadvertently.

File: lib/spack/spack/cmd/mirror.py
Comment (tgamblin) [inline lib/spack/spack/cmd/mirror.py]: same here -- default to public, make private the special case so that it is not easy to accidentally violate a license.

File: lib/spack/spack/package.py
Comment (tgamblin) [inline lib/spack/spack/package.py]: Can we make these two attributes into a directive, e.g.:

```python
class MyPackage(Package):
    redistribute(source=True, binary=False)
```

Where the defaults are (for now) `True` and `True` but eventually (after this PR) taken from license metadata?

I could see more conditions being added here and it's an easy way to allow this to evolve without adding a lot of attributes.

**Review by tgamblin — CHANGES_REQUESTED** [review]

**Comment by tgamblin** [comment]
Ok based on the comments here I have a proposal (see review comments above).  I like the idea that we will want to be more explicit for some packages, and not just rely on the license metadata.  So, keeping the ability to set this explicitly in the package seems good.

I think it should be a directive:

```python
class MyPackage(Package):
    redistribute(source=False, binary=False)
```

and once we have nixpkgs-like license metadata, we can take the defaults for that directive directly from the license information.  Until then, packagers can set it explicitly.

Sound good?

On SPDX identifiers: I think we should use them when they're available, but the intent was never to restrict the licenses we support to SPDX.  If a package has a special license, we can provide a short identifier for it and provide metadata about that in the top-level licenses file (including whether we can redistribute).  That doesn't preclude having an explicit way to override in the package, though.  I could see use cases for people marking particular packages as non-redistributable regardless of their license, so I think this will be useful anyway.

@rscohn2 does that address your concerns?  @scheibelp?

**Comment by rscohn2** [comment]
> @rscohn2 does that address your concerns?

Yes, it is a good solution.

**Comment by scheibelp** [comment]
> does that address your concerns? @scheibelp?

Yes (and the requests all make sense)

**Comment by scheibelp** [comment]
@tgamblin

> I think we should instead default to public and allow a flag for --private.

done

> Can we make these two attributes into a directive, e.g.:

done

Additionally, I ended up refactoring the logic that calculates the set of specs to mirror (before this PR, that was split across 3 functions which were all doing the same thing).

File: lib/spack/spack/cmd/buildcache.py
Comment (tgamblin) [inline lib/spack/spack/cmd/buildcache.py]:
```suggestion
        help="for a private mirror, include non-redistributable packages",
```

File: lib/spack/spack/cmd/buildcache.py
Comment (tgamblin) [inline lib/spack/spack/cmd/buildcache.py]: This shouldn't mention `redistribute_binary` anymore:

```suggestion
            f" because they cannot be redistributed: {', '.join(s.name for s in removed_specs)}\n"
            "You can use `--private` to include them."
```

Also consider using `colify()`  for the list here.

File: lib/spack/spack/cmd/mirror.py
Comment (tgamblin) [inline lib/spack/spack/cmd/mirror.py]:
```suggestion
        help="for a private mirror, include non-redistributable packages",
```

File: lib/spack/spack/directives.py
Comment (tgamblin) [inline lib/spack/spack/directives.py]: Ok sorry I should've thought about this on the first review, but I am *positive* that people are going to want to do this conditionally (another good reason for a directive).  There are packages where, depending on the version, or some variants, the package may be redistributable or not.  Also note that `license()` supports `when=`, and we'll want to set this depending on the license.

So:

```suggestion
def redistribute(source: bool = True, binary: bool = True, *, when: WhenType = None):
```

File: lib/spack/spack/directives.py
Comment (tgamblin) [inline lib/spack/spack/directives.py]: I think these should be stored in a when-keyed dictionary, e.g.:

```python
@directive("redistribute")
def redistribute(source=True, binary=True, when: WhenType = None):
    when_spec = _make_when_spec(when)
    if not when_spec:
        return

    redist = self.redistribute.setdefault(when_spec, {})
    redist["source"] = source
    redist["binary"] = binary
```

note that this doesn't affect the solve -- we should just use it after the solve to determine if a concrete spec is redistributable.

I actually *do* think something like this:

```python
redistribute: {
    "source": true,
    "binary": true,
}
```

should go on the spec itself, but I can add that part in #32312 (where I think it makes more sense)

File: lib/spack/spack/package_base.py
Comment (tgamblin) [inline lib/spack/spack/package_base.py]: this class should be unnecessary if you use a directive with a `redistribute` dict as above

**Review by tgamblin — CHANGES_REQUESTED** [review]
a few tweaks to help text and one more request for the directive.

File: lib/spack/spack/directives.py
Comment (scheibelp) [reply lib/spack/spack/directives.py]: > note that this doesn't affect the solve -- we should just use it after the solve to determine if a concrete spec is redistributable.

Currently there are spots where we don't want to concretize specs (e.g. when we do `spack mirror --all`). My concern with any `when=` approach is that - unless the `when` is strictly limited to versions - not all information will be available at mirror creation time.

We could also detect conditional redistribution settings, conservatively return `False` for non-concrete specs (thereby sacrificing the possibility of including them in a full mirror) and allowing a more fine-grained check when the associated Spec is concrete.

Comment (tgamblin) [reply lib/spack/spack/directives.py]: > My concern with any when= approach is that - unless the when is strictly limited to versions - not all information will be available at mirror creation time.

Yeah this is a good point.  I may be thinking about this too quickly, but I would say that *if* the user provides the `source=` argument, you could require the `when` spec to be either `True`, `False`, or a constraint only on version.  e.g., this would be ok:

```python
    redistribute(source=False, binary=False, when="@:3")
    redistribute(binary=False, when="+option")
```

but not this, for the reason you described:
```python
    redistribute(source=False, binary=False, when="@:3")
    redistribute(source=False, binary=False, when="+option")
```

I agree -- almost all we know at source mirror generation time the version of the tarball. 

It occurs to me that we currently don't have any redistribution information for specific patches (which can be conditional on a variant, which means you might know that *they* shouldn't be redistributed when something other than a version is constrained), but I also think every patch is at least somewhat dependent on a version, and we should leave patches for some future PR 😅.

File: lib/spack/spack/package_base.py
Comment (scheibelp) [reply lib/spack/spack/package_base.py]: Regardless of how the `@directive` is used, functionality will be required to sift through the aggregated `when` specs and determine if a given spec is redistributable. In light of https://github.com/spack/spack/pull/20185#discussion_r1545192492, the associated methods have changed, but the class is still useful.

File: lib/spack/spack/directives.py
Comment (scheibelp) [reply lib/spack/spack/directives.py]: Regarding `@directive("redistribute")`

I think it would be best to use `@directive(dicts=())` because I don't think the following makes sense:

```
redistribute(source=False, when=@2:4)
redistribute(source=True, when=@3.1)
```

I don't think we should have to offer semantics for competing enabling/disabling of redistribution, and for that reason I think users should only allowed to be set `source/binary=False` (never True); as a consequence, there is no need for a dict to store this information (all we need is a list of when specs that disable source/binary distribution).

In any case, the part about accepting a `when_spec` is handled (but not yet tested, so for now I have marked this PR as a draft).

Comment (tgamblin) [reply lib/spack/spack/directives.py]: @scheibelp I want the dictionaries to be consistent and indexed by `when` spec so that we can implement consistent logic in directives like `drop()` (see #35045). Before #40326, we had a bunch of different directive dictionary structures, which made it very hard to understand which dictionaries were structured which way.  It's better to be consistent on this than to have to implement different algorithms for filtering different dicts.

If you want to simplify the above, one option would be to do what the `license()` directive already does, which is to detect when ranges are overlapping and disallow it. Maybe generalizing [the logic here](https://github.com/spack/spack/pull/39346/files#diff-5a5ca8628a0ed62863ebe23f74fa01a6152c018d7b6693cea4f57857db22ed62R872-R885) makes sense?

Comment (scheibelp) [reply lib/spack/spack/directives.py]: > I want the dictionaries to be consistent and indexed by when spec

OK: that is done now, it looked like the values can be anything, so I made a new `DisableRedistribute` tuple

> If you want to simplify the above, one option would be to do what the license() directive already does, which is to detect when ranges are overlapping and disallow it. 

For enable/disable conflicts, these are resolved by restricting the user to disable. For disable/disable conflicts, I'm not sure if we should actually consider them conflicts in general:

```
class TestPackage():
  redistribute(source=False, when="@1.1")
  redistribute(binary=False, when="@1.1")
  redistribute(binary=False, when=@1.0:1.2+foo)
```

the first two items can conflict, but I think it's worth allowing the 2nd and 3rd together.

If you agree, then the conflict checking is just key collision.

File: lib/spack/spack/directives.py
Comment (tgamblin) [inline lib/spack/spack/directives.py:635]: Why couldn't you have:

```python
class BuiltinPackage:
    redistribute(source=False, binary=False)
```

and:

```python
class MyRepoPackage(BuiltinPackage):
    redistribute(source=True, binary=True). # override setting in superclass
```

Even within `builtin`, I could see people wanting to create derived classes depending on a license change (e.g., for [Valkey](https://www.linuxfoundation.org/press/linux-foundation-launches-open-source-valkey-community) or [OpenTofu](https://www.linuxfoundation.org/press/announcing-opentofu), which are open source forks of recently-made-closed-source products).

Maybe I'm not completely seeing what the "only allow setting to false" thing buys us here.

**Review by tgamblin — CHANGES_REQUESTED** [review]

File: lib/spack/spack/directives.py
Comment (scheibelp) [reply lib/spack/spack/directives.py:635]: Within a single package, I think we agreed that we don't want to allow competing overlapping specs like

```
redistribute(source=False, when=@2:3)
redistribute(source=True, when=@2.5)
```

given that we default to `source/binary=True`, only `False` should be allowed (`True` would only ever be needed for creating overlaps). 

You are highlighting a case of 100% overlap (i.e. key-replacement); while I agree that should be doable, for subclasses, we (will) have `drop` (which is why this info was reorganized into a Spec-keyed dictionary): I think it's not worth complicating the single-class case to offer a limited version of what `drop` will achieve in a more-straightforward manner.

Comment (tgamblin) [reply lib/spack/spack/directives.py:635]: Ok, I think this is good for now. I am not convinced that we won't want to change this, but with the `redistribute()` directive we have the interface, and the internals can change underneath.

**Review by tgamblin — APPROVED** [review]

**Pull request merged**
